O3-310: allow Put operations on an application-writable config.json file to openmrs/data/frontends/config.json

jnsereko opened the pull request:
There should be a `config.json` file that the application is allowed to manage, which lives in the application data directory.

module-spa should have a REST endpoint `PUT /ws/frontend/config` that sets the value of this file.

The file is served the usual way along with everything else in the `frontend/` directory.

The Implementer Tools should have a button that updates that file with the temporary config.

cc @dkayiwa @hadijahkyampeire 

Issue Link:  https://issues.openmrs.org/browse/O3-310

## Screenshots


Uploading screen recording.mov…



https://github.com/openmrs/openmrs-module-spa/assets/58003327/47f704b0-b967-4f8f-9dbe-9f498bc233d0

jnsereko commented on omod/src/main/java/org/openmrs/module/spa/controllers/TrialConfigController.java:
```diff
@@ -0,0 +1,24 @@
+
+@Controller
+@RequestMapping("frontend")
+public class TrialConfigController {
```
this is just a simple hack to try out my stuff.  Kindly ignore

jnsereko commented on omod/src/main/java/org/openmrs/module/spa/filter/SpaFilter.java:
```diff
@@ -28,40 +28,45 @@
 		}
-		
+
+		log.warn("Frontend request url is " + requestURL + " hence working fine. ..... spa filter");
```
this is just a simple hack to try out my stuff. Kindly ignore

jnsereko commented on omod/src/main/java/org/openmrs/module/spa/servlet/SpaServlet.java:
```diff
@@ -55,10 +82,29 @@ protected long getLastModified(HttpServletRequest servletRequest) {
-        if (SpaModuleUtils.isRemoteAssetsEnabled()) {
-            handleRemoteAssets(request, response);
+        String requestURI = request.getRequestURI();
+        if (requestURI.contains("/config")) {
```
by default, all `GET` requests return an `index.htm` file by default.  This is why i think i manually need to call the corresponding controllers

jnsereko replied:
An all my registered controllers are actually not automatically not recognised

jnsereko commented on omod/src/test/java/org/openmrs/module/spa/servlet/SpaServletTest.java:
```diff
@@ -1,98 +1,98 @@
+//
+//@ContextConfiguration(locations = {"classpath:applicationContext-service.xml",
+//        "classpath:webModuleApplicationContext.xml"}, inheritLocations = false)
+//public class SpaServletTest extends BaseModuleContextSensitiveTest {
```
pausing tests so that compile runs faster on prototyping

jnsereko commented on omod/src/test/java/org/openmrs/module/spa/servlet/SpaServletWebTest.java:
```diff
@@ -1,75 +1,75 @@
+//
+//@ContextConfiguration(locations = { "classpath:applicationContext-service.xml",
+//        "classpath:webModuleApplicationContext.xml"}, inheritLocations = false)
+//public class SpaServletWebTest extends BaseModuleContextSensitiveTest {
```
pausing tests so that compile runs faster on prototyping

jnsereko reviewed (commented):
By default, every request to `spa/` transformed with an a retrieve the `index.html` file.  
Ideally, i think, this is the reason the controllers i registered with https://github.com/openmrs/openmrs-module-spa/pull/57/commits/70e0cddc4ad72dfb475522dcb29fbe1a9c6d1fa2 were actually not being called (silent).  

I tried to call them manually but things just got messed up the more, leaving me with implementing the logic in the `doGet` and `doPut`.

I also didn't want to reimplement the existing logic of rendering the `index.html` logic 

cc @dkayiwa @ibacher @hadijahkyampeire @vasharma05 @brandones @denniskigen

ibacher commented on omod/src/main/java/org/openmrs/module/spa/servlet/SpaServlet.java:
```diff
@@ -55,10 +64,57 @@ protected long getLastModified(HttpServletRequest servletRequest) {
+            }
+        }
+    }
+    @Override
```
```suggestion

    @Override
```

ibacher commented on omod/src/main/java/org/openmrs/module/spa/servlet/SpaServlet.java:
```diff
@@ -55,10 +64,57 @@ protected long getLastModified(HttpServletRequest servletRequest) {
+            File configFile = getJsonConfigFile();
+            if (configFile.exists()) {
+                response.setContentType("application/json");
+                response.setHeader("Content-Disposition", "attachment; filename=\"" + configFile.getAbsolutePath() + "\"");
```
We do not need to set this header. It does the wrong thing.

ibacher replied:
On the other hand, we probably _do_ want to use many of the other headers set in `handleLocalAssets`

ibacher commented on omod/src/main/java/org/openmrs/module/spa/servlet/SpaServlet.java:
```diff
@@ -55,10 +64,57 @@ protected long getLastModified(HttpServletRequest servletRequest) {
+                response.setContentType("application/json");
+                response.setHeader("Content-Disposition", "attachment; filename=\"" + configFile.getAbsolutePath() + "\"");
+
+                FileInputStream input = new FileInputStream(configFile);
```
We have a utility to copy from one InputStream to an Output stream that's part of OpenmrsUtils. Please use that.

ibacher commented on omod/src/main/java/org/openmrs/module/spa/servlet/SpaServlet.java:
```diff
@@ -55,10 +64,57 @@ protected long getLastModified(HttpServletRequest servletRequest) {
-        if (SpaModuleUtils.isRemoteAssetsEnabled()) {
-            handleRemoteAssets(request, response);
+        String requestURI = request.getRequestURI();
+        if (requestURI != null && requestURI.contains("frontend/config")) {
```
I don't love using `contains` here... We should be using `endsWith()` or similar. Also... why don't we have `.json` on the end of the file path?

ibacher commented on omod/src/main/java/org/openmrs/module/spa/servlet/SpaServlet.java:
```diff
@@ -55,10 +64,57 @@ protected long getLastModified(HttpServletRequest servletRequest) {
+        String requestURI = request.getRequestURI();
+        if (requestURI.contains("frontend/config")) {
+
+            BufferedReader reader = request.getReader();
+            StringBuilder stringBuilder = new StringBuilder();
+            String line;
+
+            while ((line = reader.readLine()) != null) {
+                stringBuilder.append(line);
+            }
+
+            String requestBody = stringBuilder.toString();
+
+            File jsonConfigFile = getJsonConfigFile();
+            FileWriter writer = new FileWriter(jsonConfigFile);
+            writer.write(requestBody);
+            writer.close();
```
Again, please just use the built in method for copying from one stream to another... We can just go from the reader to the writer... There's no reason to cache the whole thing in memory.

ibacher replied:
Also, **never** close an InputStream or Writer without ensuring that it's done in a `finally` block or using the try-with-resources syntax (which does the same thing). If an exception occurs in the `writer.write()` call, the `close()` call will never be hit, which results in the application continuing to hold onto the underlying file descriptor for longer than it should.

ibacher commented on omod/src/main/java/org/openmrs/module/spa/servlet/SpaServlet.java:
```diff
@@ -55,10 +64,57 @@ protected long getLastModified(HttpServletRequest servletRequest) {
+        }
+    }
+    @Override
+    protected void doPut(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
```
From an API perspective, this is much more similar to a POST request and should probably use that verb.

ibacher replied:
There probably need to be some privilege controls around who can write to this file...

ibacher commented on omod/src/main/resources/config.xml:
```diff
@@ -39,6 +39,18 @@
     </globalProperty>
 
+	<globalProperty>
+		<property>spa.config.directory</property>
```
Why do we need a directory for the config if we're really just managing a single file?

ibacher commented on omod/src/main/java/org/openmrs/module/spa/servlet/SpaServlet.java:
```diff
@@ -178,4 +234,24 @@ private void addCacheControlHeader(HttpServletRequest request, HttpServletRespon
+        String configFolderName = as.getGlobalProperty(CONFIG_DIRECTORY, DEFAULT_CONFIG_DIRECTORY);
+        String jsonConfigFileName = as.getGlobalProperty(JSON_CONFIG_FILE_NAME, DEFAULT_JSON_CONFIG_FILE_NAME);
+
+        File configDir = new File(folder +"/"+ configFolderName);
+        if (!configDir.exists()) {
+            configDir.mkdirs(); // create directory and all parent directories if necessary
+        }
```
This seems like something we should only do when actually writing the file... Currently, we're also creating directories on a read?

ibacher commented on omod/src/main/java/org/openmrs/module/spa/servlet/SpaServlet.java:
```diff
@@ -178,4 +234,24 @@ private void addCacheControlHeader(HttpServletRequest request, HttpServletRespon
+        String configFolderName = as.getGlobalProperty(CONFIG_DIRECTORY, DEFAULT_CONFIG_DIRECTORY);
+        String jsonConfigFileName = as.getGlobalProperty(JSON_CONFIG_FILE_NAME, DEFAULT_JSON_CONFIG_FILE_NAME);
+
+        File configDir = new File(folder +"/"+ configFolderName);
```
```suggestion
        File configDir = new File(folder,  configFolderName);
```

ibacher commented on omod/src/main/java/org/openmrs/module/spa/servlet/SpaServlet.java:
```diff
@@ -55,10 +64,57 @@ protected long getLastModified(HttpServletRequest servletRequest) {
+            if (jsonConfigFile.exists()) {
+                response.setStatus(HttpServletResponse.SC_OK);
+            } else {
+                response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
```
It would be a good idea to log something and return some kind of message about what failed...

ibacher commented on omod/src/main/java/org/openmrs/module/spa/servlet/SpaServlet.java:
```diff
@@ -55,10 +64,57 @@ protected long getLastModified(HttpServletRequest servletRequest) {
+            // get the config.json file
+            File configFile = getJsonConfigFile();
+            if (configFile.exists()) {
+                response.setContentType("application/json");
```
It would be a good idea to have this set as:

```
application/json; charset=UTF-8
```

ibacher commented on omod/src/main/java/org/openmrs/module/spa/servlet/SpaServlet.java:
```diff
@@ -55,10 +64,57 @@ protected long getLastModified(HttpServletRequest servletRequest) {
+        String requestURI = request.getRequestURI();
+        if (requestURI.contains("frontend/config")) {
+
+            BufferedReader reader = request.getReader();
```
I would make some effort to validate that the things that we get passed here is actually a valid JSON document before we save it to a file...

ibacher reviewed (changes requested):
@jnsereko Thanks for the PR! I think there are some decent ideas here, but some things that need to be addressed before this can be merged in.

I think there are some design questions here that need to be handled as well, specifically around how this is supposed to interact with the implementer tools. What should happen when, having created a configuration, I go back into the implementer tools to configure a different component? What if I decide my configuration is broken and I want to get rid of it?

Finally, why is this defaulting to `/openmrs/spa/frontend/config` instead of, e.g., /openmrs/spa/config.json`?

ibacher commented:
PS The requests for `/openmrs/spa` get sent to a filter (which will run before any controllers) which forward the request to the servlet, which is why your controllers weren't getting called.

jnsereko commented:
@ibacher @dkayiwa, i think we have to change the endpoint simply to `config.json` than the earlier `frontend/config.json` so as to simplify this requirement 
By default the spa module tries to get the url as a file under the `frontend` directory.  So a request like `/openmrs/spa/config.json` will be resolved as `/openmrs/data/frontend/config.json` while a request like `/openmrs/spa/frontend/config.json` will be transformed to `/openmrs/data/frontend/frontend/config.json`

I have also removed creation of the `config` directory because it seams useless

jnsereko commented:
@dkayiwa @ibacher kindly review

dkayiwa commented on omod/src/main/java/org/openmrs/module/spa/SpaConstants.java:
```diff
@@ -28,5 +28,11 @@ private SpaConstants() {}
 	public static final String GP_REMOTE_URL = "spa.remote.url";
+
+	public static final String DEFAULT_CONFIG_DIRECTORY = "config";
+	public static final String CONFIG_DIRECTORY = "spa.config.directory";
```
Can you follow the same naming convention as for ones above?

jnsereko commented:
hey @ibacher @dkayiwa, is this fit for merge?

dkayiwa commented on omod/pom.xml:
```diff
@@ -44,6 +44,11 @@
 			<scope>provided</scope>
 		</dependency>
+		<dependency>
+			<groupId>commons-codec</groupId>
```
Could there be a reason why you did not avoid adding this dependency by simply using `java.util.Base64`?

dkayiwa commented on omod/src/main/java/org/openmrs/module/spa/SpaConstants.java:
```diff
@@ -28,5 +28,11 @@ private SpaConstants() {}
+	public static final String CONFIG_DIRECTORY = "spa.config.directory";
+
+	public static  final String DEFAULT_JSON_CONFIG_FILE_NAME = "config.json";
+	public static  final String GP_JSON_CONFIG_FILE_NAME = "spa.json.config.filename";
```
Is there any reason why you want the file name to also be configured?

dkayiwa commented on omod/src/main/java/org/openmrs/module/spa/servlet/SpaServlet.java:
```diff
@@ -62,6 +81,75 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response) t
+                if (basicAuth != null) {
+                    // check that header is in format "Basic ${base64encode(username + ":" + password)}"
+                    if (basicAuth.startsWith("Basic")) {
+                        try {
```
Can we simplify the `doPost` by wrapping this in private util method/s? This method/s could still be in the same class.

dkayiwa commented on omod/src/main/java/org/openmrs/module/spa/servlet/SpaServlet.java:
```diff
@@ -62,6 +81,75 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response) t
+                File jsonConfigFile = getJsonConfigFile();
+                try {
+                    BufferedReader reader = request.getReader();
+                    StringBuilder stringBuilder = new StringBuilder();
```
Can we wrap this too in a utility method?

dkayiwa commented:
@jnsereko are you still working on this?

jnsereko commented:
@dkayiwa  @ibacher I have cleaned the code

dkayiwa commented on omod/src/main/java/org/openmrs/module/spa/servlet/SpaServlet.java:
```diff
@@ -178,4 +275,17 @@ private void addCacheControlHeader(HttpServletRequest request, HttpServletRespon
+            throw new ModuleException("SPA frontend repository is not a directory at: " + folder.getAbsolutePath());
+        }
+
+        AdministrationService as = Context.getAdministrationService();
```
Why assign it to a variable if you are just using it once?
Besides, `Context.getAdministrationService()` is clearer than `as` 😊

jnsereko committed:
O3-310: allow Put operations on an application-writable config.json file

config.json file name shouldn't be dynamic

dynamically serve 'frontend/config.json' endpoint

remove unrequited change

jnsereko commented:
Hey @dkayiwa 
is it ok for merging

dkayiwa commented:
@jnsereko have you done another round of testing with the latest O3?

jnsereko commented:
hey @ibacher @ibacher 
Is it that 3.x has finally migrated away form using the this module?.  Deploying my changes in docker are no-longer working like it used to some time back.

ibacher commented:
> Is it that 3.x has finally migrated away form using the this module?

The Docker images have never used module-spa.

jnsereko commented:
The reason why I asked @ibacher is because deployed the spa module into
docker for 3.x but all requests to spa/config.json actually returned 404
yet it used to work sometime back (on opening this PR)

On Mon, May 29, 2023, 3:34 PM Ian ***@***.***> wrote:

> Is it that 3.x has finally migrated away form using the this module?
>
> The Docker images have never used module-spa.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openmrs/openmrs-module-spa/pull/57#issuecomment-1567084281>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AN2Q67ZXA5VF4HBKG2MWE4DXISJWZANCNFSM6AAAAAAVKTVR74>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

ibacher commented:
When you were testing this previously, I'm pretty sure you were using the SDK. We haven't changed anything about how the routing works. If you're using the Docker containers, requests to `/openmrs/spa` will never hit the backend... and that's been the case since we put the Docker containers together. (This is, in fact, what allows us to provide the "single-page" experience; any request to `/openmrs/spa` is forwarded to the frontend container and any request for a URL in the frontend container that does not otherwise resolve to something serves the single-page content).

jnsereko commented:
This now makes sense @ibacher
Maybe I was using the sdk to test those days but now using docker.

Actually logs were showed only frontend and getway images.  I was wondering
why the backed was not showing.

Since I prefer docker, and now that all requests to */spa/* go to frontend,
how can I override this?   Additionally, would there another better option?


On Mon, May 29, 2023, 7:19 PM Ian ***@***.***> wrote:

> When you were testing this previously, I'm pretty sure you were using the
> SDK. We haven't changed anything about how the routing works. If you're
> using the Docker containers, requests to /openmrs/spa will never hit the
> backend... and that's been the case since we put the Docker containers
> together. (This is, in fact, what allows us to provide the "single-page"
> experience; any request to /openmrs/spa is forwarded to the frontend
> container and any request for a URL in the frontend container that does not
> otherwise resolve to something serves the single-page content).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openmrs/openmrs-module-spa/pull/57#issuecomment-1567322538>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AN2Q675TFNAHBUMBCX3ANUTXITEBVANCNFSM6AAAAAAVKTVR74>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

jnsereko commented:
> When you were testing this previously, I'm pretty sure you were using the SDK. We haven't changed anything about how the routing works. If you're using the Docker containers, requests to `/openmrs/spa` will never hit the backend... and that's been the case since we put the Docker containers together. (This is, in fact, what allows us to provide the "single-page" experience; any request to `/openmrs/spa` is forwarded to the frontend container and any request for a URL in the frontend container that does not otherwise resolve to something serves the single-page content).

i have created a [thread ](https://talk.openmrs.org/t/moving-forward-with-post-get-on-openmrs-module-spa-in-docker-environment/39779)for how to move forward because we cant limit this feature to working only in the `openmrs-sdk` environment.

jnsereko commented:
thanks to @dkayiwa, i dynamically defined a servlet to expose an end-point `openmrs/ws/frontend/config.json` rather than using `openmrs/spa/config.json` where the request does not reach the backend in the docker environment.  

This works fine in both `openmrs-sdk` and `openmrs-docker` environment

jnsereko committed:
push to \'/ws/frontend/*\'

jnsereko commented:
@dkayiwa @ibacher the frontend implementation has been approved see https://github.com/openmrs/openmrs-esm-core/pull/629#pullrequestreview-1497579540

dkayiwa commented on omod/src/main/java/org/openmrs/module/spa/web/SpaWebComponentRegistrar.java:
```diff
@@ -0,0 +1,50 @@
+    public void setServletContext(ServletContext servletContext) {
+
+        try {
+            ServletRegistration openmrsServletReg = servletContext.getServletRegistration("openmrs");
```
What is this mapping used for?

jnsereko replied:
I thought we need to ensure that any mapping is supposed to be on top of the OpenMRS servlet.  just to eliminate circumstances like `https:localhost:port/ws/frontend/config.json`

dkayiwa replied:
What happens when you remove line 32?

jnsereko replied:
If `servletContext.getServletRegistration("openmrs");` isn't necessary @dkayiwa,  how can i add mappings to the Openmrs Servlet then?

dkayiwa replied:
What happens when you remove line 32?

jnsereko replied:
works fine.

jnsereko committed:
remove unnecessary mappings

dkayiwa commented on omod/src/main/java/org/openmrs/module/spa/servlet/SpaServlet.java:
```diff
@@ -62,6 +74,84 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response) t
+                String basicAuth = request.getHeader("Authorization");
+                if (basicAuth != null) {
+                    // check that header is in format "Basic ${base64encode(username + ":" + password)}"
+                    if (isValidAuthFormat(response, basicAuth)) return;
```
https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-CodeStyle

dkayiwa commented on omod/src/main/java/org/openmrs/module/spa/web/SpaWebComponentRegistrar.java:
```diff
@@ -0,0 +1,47 @@
+            ServletRegistration servletReg = servletContext.addServlet("spaServlet", new SpaServlet());
+            servletReg.addMapping("/ws/frontend/config.json");
+
+            Dynamic filter = servletContext.addFilter("spaFilter", new SpaFilter());
```
Is the SpaFilter part of this ticket?

dkayiwa commented on omod/src/main/java/org/openmrs/module/spa/servlet/SpaServlet.java:
```diff
@@ -62,6 +74,84 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response) t
+    @Override
+    protected void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        String requestURI = request.getRequestURI();
+        if (requestURI.endsWith("/config.json")) {
```
Can you reduce the deep nesting of these if statements?

dkayiwa commented on omod/src/main/java/org/openmrs/module/spa/servlet/SpaServlet.java:
```diff
@@ -62,6 +74,84 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response) t
+            while ((line = reader.readLine()) != null) {
+                stringBuilder.append(line);
+            }
+            String requestBody = stringBuilder.toString();
```
I do not see much value in the requestBody variable.

brandones commented:
@jnsereko Do you still have bandwidth to work on this? Could you please work with @dkayiwa to get it across the finish line? It's soooooo close!!

dkayiwa commented:
I completely agree! 👍

jnsereko commented:
updating it now

jnsereko committed:
O3-310: fix violated java conventions

dkayiwa commented:
Is this functionality going to be only for those who are not using the docker setup of O3 (Because it does not load the spa module)

jnsereko commented:
> Is this functionality going to be only for those who are not using the docker setup of O3 (Because it does not load the spa module)

We can add the module to distro if we choose to.  I can do that with your permission

dkayiwa commented:
I do not think that would be the best option. One other option is adding this to the rest webservices module. This calls for some sort of discussion on talk.

jnsereko commented:
> I do not think that would be the best option. One other option is adding this to the rest webservices module. This calls for some sort of discussion on talk.

I have created a thread: https://talk.openmrs.org/t/adding-openmrs-spa-module-to-dev3-vs-migrating-some-of-its-logic-to-rest-webservices/41250

jnsereko commented:
@dkayiwa PR created in Rest web services repo https://github.com/openmrs/openmrs-module-webservices.rest/pull/595
